File: SortVision/public/code/heap/lua/heapSort.lua
-- Heap Sort (in-place, max-heap)
-- Time: Best/Avg/Worst O(n log n), Space: O(1)
local function heapify(arr, n, i)
  local largest = i
  local left = 2 * i
  local right = 2 * i + 1

  if left <= n and arr[left] > arr[largest] then largest = left end
  if right <= n and arr[right] > arr[largest] then largest = right end

  if largest ~= i then
    arr[i], arr[largest] = arr[largest], arr[i]
    heapify(arr, n, largest)
  end
end

local function heapSort(arr)
  local n = #arr
  for i = math.floor(n / 2), 1, -1 do heapify(arr, n, i) end
  for i = n, 2, -1 do
    arr[1], arr[i] = arr[i], arr[1]
    heapify(arr, i - 1, 1)
  end
  return arr
end

local arr = {64, 25, 12, 22, 11}
print(table.concat(heapSort(arr), ", "))

File: SortVision/public/code/heap/r/heapSort.r
# Heap Sort
# Time: Best/Avg/Worst O(n log n), Space: O(1)
heapify <- function(arr, n, i) {
  largest <- i
  left <- 2 * i
  right <- 2 * i + 1

  if (left <= n && arr[left] > arr[largest]) largest <- left
  if (right <= n && arr[right] > arr[largest]) largest <- right

  if (largest != i) {
    temp <- arr[i]
    arr[i] <- arr[largest]
    arr[largest] <- temp
    arr <- heapify(arr, n, largest)
  }
  arr
}

heap_sort <- function(arr) {
  n <- length(arr)

  for (i in floor(n / 2):1) arr <- heapify(arr, n, i)
  for (i in n:2) {
    temp <- arr[1]
    arr[1] <- arr[i]
    arr[i] <- temp
    arr <- heapify(arr, i - 1, 1)
  }

  arr
}

arr <- c(64, 25, 12, 22, 11)
print(heap_sort(arr))
